README.md
[![Build Status](https://travis-ci.org/osstech-jp/akerun-sum.svg?branch=master)](https://travis-ci.org/osstech-jp/akerun-sum)

### 入退出集計プログラム
NFCカードのドアキー[アケルン](https://akerun.com/)の入退出記録から、勤務日数や勤務時間を集計するプログラムです。

### 以下の環境で動作確認
  * Windows 10 Home, Python 3.4.3
  * Ubuntu 16.04.2 LTS, Python 3.5.2

### 使用方法
`akerun-sum.py -i inputfile -o outputfile -d yyyymm [-f n]`  
##### 引数
    -i  入力ファイル名
    -o  出力ファイル名
    -d  集計期間 yyyymm の形式で指定
    -f  出力タイプ 初期値は0
        0  出力パターン1
        1  出力パターン2

### 実行例

    akerun-sum.py -i input-euc.csv -o output-euc.csv -d 201610
    akerun-sum.py -i input-anotherformat.csv -o output-anotherformat.csv -d 201610　-f 1

社員数やレコードの数はリストで管理しているため無制限


### 想定している入力ファイル
DATE,AKERUN,USER,LOCK,CLIENTのカラムを持つCSVファイル

##### DATE
日付データ  
`yyyy/mm/dd hh:mm`と`yyyy-mm-dd hh:mm:ss`の2パターンに対応  
昇順にソートされていることが前提
##### AKERUN
本プログラムでは使用していない
##### USER
社員名データ
##### LOCK
* 入室：オフィスに入った
* 退室：オフィスから出た
* 解錠：オフィスに入室したか退室のどちらか
* 施錠：鍵を締めた（本プログラムでは使用していない）

##### CLIENT
鍵の種類（本プログラムでは使用していない）

### 出力ファイル
出力ファイルは2パターンあり、引数によって切替可能  
文字コードは入力ファイルに合わせる

##### 出力パターン1
Excelファイルで開くことを想定

|氏名|就業日数|就業時間|yyyy/mm/dd入|yyyy/mm/dd退|yyyy/mm/dd時|…|
|:-:|:-:|:-:|:-:|:-:|:-:|:-:|
|山田太郎|2|13.5|8:47|10:12|1.25|…|
|山田次郎|2|20.5|8:47|20:12|11.25|…|
|:|:|:|:|:|:|:|

##### 出力パターン2
通常のCSVファイル

|||||
|:-:|:-:|:-:|:-:|
|氏名|山田太郎|||
|集計期間|yyyymm|||
|就業日数|2|||
|就業時間|13.5|||
|月日|入室時刻|退室時刻|就業時間|
|yyyy/mm/dd|8:47|10:12|1.25|
|:|:|:|:|
|||||
|氏名|山田次郎|||
|集計期間|yyyymm|||
|就業日数|2|||
|就業時間|13.5|||
|月日|入室時刻|退室時刻|就業時間|
|yyyy/mm/dd|8:47|20:12|11.25|
|:|:|:|:|


changelog.rst
更新履歴
========================

* v1.2.2

  - オプションパーサーにClickを利用
  - pyproject.tomlを利用

* v1.2.1

  - Akerunデータのヘッダーが日本語になったのでそれに合わせて修正


test.sh
#! /bin/bash

# test tool

VENV=testenv
PYTHON3=python3
SCRIPT=./${VENV}/bin/akerun-sum

diff_outputfile() {
  outfile=$1
  orgfile=$2
  inputfile=$3
  errmasage="err when used [$3] and [$2]"
  diff $outfile $orgfile > log
  if [ $? -ne 0 ]; then
    echo $errmasage
    exit $?;
  else
    echo "$inputfile Success"
    rm log
  fi
}

init() {
  python3 -m venv ${VENV}
  ./${VENV}/bin/python -m pip install -U pip setuptools wheel
  ./${VENV}/bin/python -m pip install .
}

init

${SCRIPT} -i test/input-utf8.csv -o output.csv -d 201610 -f 1 || exit $?
diff_outputfile output.csv test/output-utf8.csv test/input-utf8.csv

${SCRIPT} -i test/input-sjis.csv -o output.csv -d 201610 -f 1 || exit $?
diff_outputfile output.csv test/output-sjis.csv test/input-sjis.csv

${SCRIPT} -i test/input-euc.csv -o output.csv -d 201610 -f 1 || exit $?
diff_outputfile output.csv test/output-euc.csv test/input-euc.csv

${SCRIPT} -i test/input-anotherdate.csv -o output.csv -d 201702 -f 1 || exit $?
diff_outputfile output.csv test/output-anotherdate.csv test/input-anotherdate.csv

${SCRIPT} -i test/input-anotherdate.csv -o output.csv -d 201703 -f 1 || exit $?
diff_outputfile output.csv test/output-empty.csv test/input-anotherdate.csv

${SCRIPT} -i test/input-anotherformat.csv -o output.csv -d 201610 -f 0 || exit $?
diff_outputfile output.csv test/output-anotherformat.csv test/input-anotherformat.csv

${SCRIPT} -i test/input-anotherformat.csv -o output.csv -d 201610 || exit $?
diff_outputfile output.csv test/output-anotherformat.csv test/input-anotherformat.csv

rm output.csv
rm -r ${VENV}


test/run.sh
#! /bin/sh

python3 ../akerun-sum.py -i input-utf8.csv -o output.csv -d 201610
python3 ../akerun-sum.py -i input-sjis.csv -o output.csv -d 201610
python3 ../akerun-sum.py -i input-euc.csv -o output.csv -d 201610



.travis.sh
#! /bin/bash

# test tool

VENV=testenv
PYTHON3=python3
SCRIPT=akerun-sum

diff_outputfile() {
  outfile=$1
  orgfile=$2
  inputfile=$3
  errmasage="err when used [$3] and [$2]"
  diff $outfile $orgfile > log
  if [ $? -ne 0 ]; then
    echo $errmasage
    exit $?;
  else
    rm log
  fi
}

init() {
  python3 -m pip install .
}

init

python3 -m pip list
which akerun-sum

${SCRIPT} -i test/input-utf8.csv -o output.csv -d 201610 -f 1 || exit $?
diff_outputfile output.csv test/output-utf8.csv test/input-utf8.csv

${SCRIPT} -i test/input-sjis.csv -o output.csv -d 201610 -f 1 || exit $?
diff_outputfile output.csv test/output-sjis.csv test/input-sjis.csv

${SCRIPT} -i test/input-euc.csv -o output.csv -d 201610 -f 1 || exit $?
diff_outputfile output.csv test/output-euc.csv test/input-euc.csv

${SCRIPT} -i test/input-anotherdate.csv -o output.csv -d 201702 -f 1 || exit $?
diff_outputfile output.csv test/output-anotherdate.csv test/input-anotherdate.csv

${SCRIPT} -i test/input-anotherdate.csv -o output.csv -d 201703 -f 1 || exit $?
diff_outputfile output.csv test/output-empty.csv test/input-anotherdate.csv

${SCRIPT} -i test/input-anotherformat.csv -o output.csv -d 201610 -f 0 || exit $?
diff_outputfile output.csv test/output-anotherformat.csv test/input-anotherformat.csv

${SCRIPT} -i test/input-anotherformat.csv -o output.csv -d 201610 || exit $?
diff_outputfile output.csv test/output-anotherformat.csv test/input-anotherformat.csv

rm output.csv


akerun-sum.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*- vim:shiftwidth=4:expandtab:

import math
import calendar
import datetime
import codecs
import csv
import sys
import os

DAYSTART = '0300'
ROUNDDOWNTIME = 15


def option_parser():
    usage = 'Usage: python {} \
  -i <input filename> -o <output filename> \
  -d <period(yyyymm)> [-f formatnumber]'\
             .format(__file__)
    arguments = sys.argv
    if len(arguments) != 7 and len(arguments) != 9:
        print(usage)
        sys.exit()

    index_number = [1, 3, 5]
    if len(arguments) == 9:
        index_number = [1, 3, 5, 7]

    # default
    format_num = 0
    for index in index_number:
        option = arguments[index]
        if option == '-i':
            input_filename = arguments[index + 1]
        elif option == '-o':
            output_filename = arguments[index + 1]
        elif option == '-d':
            period = arguments[index + 1]
        elif option == '-f':
            format_num = int(arguments[index + 1])
            if format_num not in [0, 1]:
                print('This format number is not exist.')
                sys.exit()
        else:
            print(usage)
            sys.exit()

    return {'input_filename': input_filename,
            'output_filename': output_filename,
            'period': period,
            'format_num': format_num}


def input_data(filename):
    lookup = ('utf_8', 'euc_jp', 'euc_jis_2004', 'euc_jisx0213',
              'shift_jis', 'shift_jis_2004', 'shift_jisx0213',
              'iso2022jp', 'iso2022_jp_1', 'iso2022_jp_2', 'iso2022_jp_3',
              'iso2022_jp_ext', 'latin_1', 'ascii')
    encode = None
    data_list = []

    for encoding in lookup:
        try:
            f = codecs.open(filename, 'r', encoding)
            encode = encoding
            reader = csv.reader(f)

            for row in reader:
                date = row[0]
                user = row[2]
                lock = row[3]
                data = {'date': date, 'user': user, 'lock': lock}
                data_list.append(data)
            break
        except:
            data_list.clear()
            pass
    if isinstance(encode, str):
        return data_list, encode
    else:
        raise LookupError


def data_shaping(data_list, period):
    # 2:00 -> 23:00
    date_format_list = ['%Y/%m/%d %H:%M', '%Y-%m-%d %H:%M:%S']
    day_start = datetime.datetime.strptime(DAYSTART, '%H%M')
    for data in data_list:
        for date_format in date_format_list:
            try:
                date = datetime.datetime.strptime(data['date'], date_format)
                break
            except:
                pass
        data['date'] = date
        data['date'] -= datetime.timedelta(hours=day_start.hour)
        data['date'] -= datetime.timedelta(minutes=day_start.minute)

    # data mining
    period_start = datetime.datetime.strptime(period, '%Y%m')
    day_range = calendar.monthrange(period_start.year, period_start.month)[1]
    period_end = period_start + datetime.timedelta(days=day_range)

    mining_data = []
    user_list = []
    shaped_data = []
    for data in data_list:
        if period_start <= data['date'] and data['date'] < period_end\
                and data['lock'] in ['入室', '退室', '解錠']\
                and data['user'] != '':
            mining_data.append(data)
            if data['user'] not in user_list:
                user_list.append(data['user'])
                shaped_data.append({
                    'name': data['user'], 'timecard_data': [],
                    'writed_days': [], 'period': period})

    # data reconstruction
    for data in mining_data:
        index = user_list.index(data['user'])

        if data['date'].day not in shaped_data[index]['writed_days']:
            shaped_data[index]['timecard_data'].append({'day': data['date'].day})
            shaped_data[index]['writed_days'].append(data['date'].day)

        timecard_data_index = shaped_data[index]['writed_days'].index(data['date'].day)

        if data['lock'] == '入室':
            if 'in_time' not in shaped_data[index]['timecard_data'][timecard_data_index]:
                shaped_data[index]['timecard_data'][timecard_data_index]['in_time'] = data['date']

        elif data['lock'] == '退室':
            shaped_data[index]['timecard_data'][timecard_data_index]['out_time'] = data['date']

        elif data['lock'] == '解錠':
            if 'in_time' not in shaped_data[index]['timecard_data'][timecard_data_index]:
                shaped_data[index]['timecard_data'][timecard_data_index]['in_time'] = data['date']
            else:
                shaped_data[index]['timecard_data'][timecard_data_index]['out_time'] = data['date']

    # data totalization
    for data in shaped_data:
        total_working_hours = 0
        total_working_days = 0

        for timecard_data in data['timecard_data']:
            if 'in_time' in timecard_data\
                    and 'out_time' in timecard_data:
                diff = timecard_data['out_time'] - timecard_data['in_time']
                if diff.days < 0:
                    timecard_data['working_hours'] = 0
                else:
                    diff_sec = diff.seconds
                    diff_min = diff_sec / 60
                    diff_min = math.floor(diff_min / ROUNDDOWNTIME) * ROUNDDOWNTIME
                    diff_hour = diff_min / 60

                    timecard_data['working_hours'] = diff_hour
                total_working_hours += diff_hour

            else:
                timecard_data['working_hours'] = 0

            total_working_days += 1

        data['total_working_hours'] = total_working_hours
        data['total_working_days'] = total_working_days

    return shaped_data


def output_data0(filename, encode, shaped_data):
    # make datelist
    day_list = []
    for data in shaped_data:
        for timecard in data['timecard_data']:
            if timecard['day'] not in day_list:
                day_list.append(timecard['day'])
    day_list.sort()

    period = shaped_data[0]['period']

    header = ['氏名', '就業日数', '就業時間']
    for day in day_list:
        day_str = period[0:4] + '/'\
            + str(int(period[4:6])) + '/'\
            + str(day)
        header.append(day_str + '入')
        header.append(day_str + '退')
        header.append(day_str + '時')

    with codecs.open(filename, 'w', encode) as f:
        writer = csv.writer(f, lineterminator=os.linesep)
        writer.writerow(header)
        for data in shaped_data:
            row = [data['name'], data['total_working_days'], data['total_working_hours']]
            curr_index = 0
            writedata_index = 0
            while curr_index < len(day_list):
                if writedata_index < len(data['timecard_data'])\
                        and data['timecard_data'][writedata_index]['day'] == day_list[curr_index]:
                    timecard = data['timecard_data'][writedata_index]
                    day_start = datetime.datetime.strptime(DAYSTART, '%H%M')
                    if 'in_time' in timecard:
                        # 23:00 -> 26:00
                        timecard['in_time'] += datetime.timedelta(minutes=day_start.minute)
                        timecard['in_time'] += datetime.timedelta(hours=day_start.hour)
                        hour = timecard['in_time'].hour
                        if hour < day_start.hour:
                            hour += 24

                        in_time_str = str(hour) + ':'\
                            + timecard['in_time'].strftime('%M')
                    else:
                        in_time_str = ''

                    if 'out_time' in timecard:
                        # 23:00 -> 26:00
                        timecard['out_time'] += datetime.timedelta(minutes=day_start.minute)
                        timecard['out_time'] += datetime.timedelta(hours=day_start.hour)
                        hour = timecard['out_time'].hour
                        if hour < day_start.hour:
                            hour += 24
                        out_time_str = str(hour) + ':'\
                            + timecard['out_time'].strftime('%M')
                    else:
                        out_time_str = ''

                    row.append(in_time_str)
                    row.append(out_time_str)
                    row.append(timecard['working_hours'])
                    writedata_index += 1
                else:
                    row.append('')
                    row.append('')
                    row.append('')
                curr_index += 1

            writer.writerow(row)
            row.clear()


def output_data1(filename, encode, shaped_data):
    with codecs.open(filename, 'w', encode) as f:
        writer = csv.writer(f, lineterminator=os.linesep)
        for data in shaped_data:
            writer.writerow(['氏名', data['name'], '', '', '', ''])
            writer.writerow(['集計期間', data['period'], '', '', '', ''])
            writer.writerow(['就業日数', data['total_working_days'], '', '', '', ''])
            writer.writerow(['就業時間', data['total_working_hours'], '', '', '', ''])
            writer.writerow(['月日', '入室時刻', '退出時刻', '就業時間', '', ''])
            for timecard in data['timecard_data']:
                date_str = data['period'][0:4] + '/'\
                    + str(int(data['period'][4:6])) + '/'\
                    + str(timecard['day'])

                day_start = datetime.datetime.strptime(DAYSTART, '%H%M')
                if 'in_time' in timecard:
                    # 23:00 -> 26:00
                    timecard['in_time'] += datetime.timedelta(minutes=day_start.minute)
                    timecard['in_time'] += datetime.timedelta(hours=day_start.hour)
                    hour = timecard['in_time'].hour
                    if hour < day_start.hour:
                        hour += 24

                    in_time_str = str(hour) + ':'\
                        + timecard['in_time'].strftime('%M')
                else:
                    in_time_str = ''

                if 'out_time' in timecard:
                    # 23:00 -> 26:00
                    timecard['out_time'] += datetime.timedelta(minutes=day_start.minute)
                    timecard['out_time'] += datetime.timedelta(hours=day_start.hour)
                    hour = timecard['out_time'].hour
                    if hour < day_start.hour:
                        hour += 24
                    out_time_str = str(hour) + ':'\
                        + timecard['out_time'].strftime('%M')
                else:
                    out_time_str = ''

                writer.writerow([
                    date_str, in_time_str, out_time_str, str(timecard['working_hours']), '', ''])

            writer.writerow(['', '', '', '', '', ''])


if __name__ == '__main__':
    commandline_vars = option_parser()
    data_list, encode = input_data(commandline_vars['input_filename'])
    shaped_data = data_shaping(data_list, commandline_vars['period'])
    if commandline_vars['format_num'] == 0:
        output_data0(commandline_vars['output_filename'], encode, shaped_data)
    elif commandline_vars['format_num'] == 1:
        output_data1(commandline_vars['output_filename'], encode, shaped_data)


akerun_sum.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*- vim:shiftwidth=4:expandtab:

import math
import calendar
import datetime
import codecs
import csv
import sys
import os

DAYSTART = '0300'
ROUNDDOWNTIME = 15

KEYS = {'date': 'DATE', 'user': 'USER', 'lock': 'LOCK'}


def option_parser():
    usage = 'Usage: python {} \
  -i <input filename> -o <output filename> \
  -d <period(yyyymm)> [-f formatnumber]'\
             .format(__file__)
    arguments = sys.argv
    if len(arguments) != 7 and len(arguments) != 9:
        print(usage)
        sys.exit()

    index_number = [1, 3, 5]
    if len(arguments) == 9:
        index_number = [1, 3, 5, 7]

    # default
    format_num = 0
    for index in index_number:
        option = arguments[index]
        if option == '-i':
            input_filename = arguments[index + 1]
        elif option == '-o':
            output_filename = arguments[index + 1]
        elif option == '-d':
            period = arguments[index + 1]
        elif option == '-f':
            format_num = int(arguments[index + 1])
            if format_num not in [0, 1]:
                print('This format number is not exist.')
                sys.exit()
        else:
            print(usage)
            sys.exit()

    return {'input_filename': input_filename,
            'output_filename': output_filename,
            'period': period,
            'format_num': format_num}


def input_data(filename):
    lookup = ('utf_8', 'euc_jp', 'euc_jis_2004', 'euc_jisx0213',
              'shift_jis', 'shift_jis_2004', 'shift_jisx0213',
              'iso2022jp', 'iso2022_jp_1', 'iso2022_jp_2', 'iso2022_jp_3',
              'iso2022_jp_ext', 'latin_1', 'ascii')
    encode = None

    for encoding in lookup:
        try:
            f = codecs.open(filename, 'r', encoding)
            encode = encoding
            reader = csv.DictReader(f)
            data_list = [d for d in reader]
            return data_list, encode
        except:
            continue
    raise LookupError


def data_shaping(data_list, period):
    # 2:00 -> 23:00
    date_format_list = ['%Y/%m/%d %H:%M', '%Y-%m-%d %H:%M:%S']
    day_start = datetime.datetime.strptime(DAYSTART, '%H%M')
    for data in data_list:
        for date_format in date_format_list:
            try:
                data[KEYS['date']] = datetime.datetime.strptime(data[KEYS['date']], date_format)
                data[KEYS['date']] -= datetime.timedelta(hours=day_start.hour)
                data[KEYS['date']] -= datetime.timedelta(minutes=day_start.minute)
                break
            except:
                pass
    # data mining
    period_start = datetime.datetime.strptime(period, '%Y%m')
    day_range = calendar.monthrange(period_start.year, period_start.month)[1]
    period_end = period_start + datetime.timedelta(days=day_range)

    mining_data = []
    user_list = []
    shaped_data = []
    for data in data_list:
        if isinstance(data[KEYS['date']], str):
            continue
        if period_start <= data[KEYS['date']] and data[KEYS['date']] < period_end\
                and data[KEYS['lock']] in ['入室', '退室', '解錠']\
                and data[KEYS['user']] != '':
            mining_data.append(data)
            if data[KEYS['user']] not in user_list:
                user_list.append(data[KEYS['user']])
                shaped_data.append({
                    'name': data[KEYS['user']], 'timecard_data': [],
                    'writed_days': [], 'period': period})

    # data reconstruction
    for data in mining_data:
        index = user_list.index(data[KEYS['user']])

        if data[KEYS['date']].day not in shaped_data[index]['writed_days']:
            shaped_data[index]['timecard_data'].append({'day': data[KEYS['date']].day})
            shaped_data[index]['writed_days'].append(data[KEYS['date']].day)

        timecard_data_index = shaped_data[index]['writed_days'].index(data[KEYS['date']].day)

        if data[KEYS['lock']] == '入室':
            if 'in_time' not in shaped_data[index]['timecard_data'][timecard_data_index]:
                shaped_data[index]['timecard_data'][timecard_data_index]['in_time'] = data[KEYS['date']]

        elif data[KEYS['lock']] == '退室':
            shaped_data[index]['timecard_data'][timecard_data_index]['out_time'] = data[KEYS['date']]

        elif data[KEYS['lock']] == '解錠':
            if 'in_time' not in shaped_data[index]['timecard_data'][timecard_data_index]:
                shaped_data[index]['timecard_data'][timecard_data_index]['in_time'] = data[KEYS['date']]
            else:
                shaped_data[index]['timecard_data'][timecard_data_index]['out_time'] = data[KEYS['date']]

    # data totalization
    for data in shaped_data:
        total_working_hours = 0
        total_working_days = 0

        for timecard_data in data['timecard_data']:
            if 'in_time' in timecard_data\
                    and 'out_time' in timecard_data:
                diff = timecard_data['out_time'] - timecard_data['in_time']
                if diff.days < 0:
                    timecard_data['working_hours'] = 0
                else:
                    diff_sec = diff.seconds
                    diff_min = diff_sec / 60
                    diff_min = math.floor(diff_min / ROUNDDOWNTIME) * ROUNDDOWNTIME
                    diff_hour = diff_min / 60

                    timecard_data['working_hours'] = diff_hour
                total_working_hours += diff_hour

            else:
                timecard_data['working_hours'] = 0

            total_working_days += 1

        data['total_working_hours'] = total_working_hours
        data['total_working_days'] = total_working_days

    return shaped_data


def output_data0(filename, encode, shaped_data):
    # make datelist
    day_list = []
    for data in shaped_data:
        for timecard in data['timecard_data']:
            if timecard['day'] not in day_list:
                day_list.append(timecard['day'])
    day_list.sort()

    period = shaped_data[0]['period']

    header = ['氏名', '就業日数', '就業時間']
    for day in day_list:
        day_str = period[0:4] + '/'\
            + str(int(period[4:6])) + '/'\
            + str(day)
        header.append(day_str + '入')
        header.append(day_str + '退')
        header.append(day_str + '時')

    with codecs.open(filename, 'w', encode) as f:
        writer = csv.writer(f, lineterminator=os.linesep)
        writer.writerow(header)
        for data in shaped_data:
            row = [data['name'], data['total_working_days'], data['total_working_hours']]
            curr_index = 0
            writedata_index = 0
            while curr_index < len(day_list):
                if writedata_index < len(data['timecard_data'])\
                        and data['timecard_data'][writedata_index]['day'] == day_list[curr_index]:
                    timecard = data['timecard_data'][writedata_index]
                    day_start = datetime.datetime.strptime(DAYSTART, '%H%M')
                    if 'in_time' in timecard:
                        # 23:00 -> 26:00
                        timecard['in_time'] += datetime.timedelta(minutes=day_start.minute)
                        timecard['in_time'] += datetime.timedelta(hours=day_start.hour)
                        hour = timecard['in_time'].hour
                        if hour < day_start.hour:
                            hour += 24

                        in_time_str = str(hour) + ':'\
                            + timecard['in_time'].strftime('%M')
                    else:
                        in_time_str = ''

                    if 'out_time' in timecard:
                        # 23:00 -> 26:00
                        timecard['out_time'] += datetime.timedelta(minutes=day_start.minute)
                        timecard['out_time'] += datetime.timedelta(hours=day_start.hour)
                        hour = timecard['out_time'].hour
                        if hour < day_start.hour:
                            hour += 24
                        out_time_str = str(hour) + ':'\
                            + timecard['out_time'].strftime('%M')
                    else:
                        out_time_str = ''

                    row.append(in_time_str)
                    row.append(out_time_str)
                    row.append(timecard['working_hours'])
                    writedata_index += 1
                else:
                    row.append('')
                    row.append('')
                    row.append('')
                curr_index += 1

            writer.writerow(row)
            row.clear()


def output_data1(filename, encode, shaped_data):
    with codecs.open(filename, 'w', encode) as f:
        writer = csv.writer(f, lineterminator=os.linesep)
        for data in shaped_data:
            writer.writerow(['氏名', data['name'], '', '', '', ''])
            writer.writerow(['集計期間', data['period'], '', '', '', ''])
            writer.writerow(['就業日数', data['total_working_days'], '', '', '', ''])
            writer.writerow(['就業時間', data['total_working_hours'], '', '', '', ''])
            writer.writerow(['月日', '入室時刻', '退出時刻', '就業時間', '', ''])
            for timecard in data['timecard_data']:
                date_str = data['period'][0:4] + '/'\
                    + str(int(data['period'][4:6])) + '/'\
                    + str(timecard['day'])

                day_start = datetime.datetime.strptime(DAYSTART, '%H%M')
                if 'in_time' in timecard:
                    # 23:00 -> 26:00
                    timecard['in_time'] += datetime.timedelta(minutes=day_start.minute)
                    timecard['in_time'] += datetime.timedelta(hours=day_start.hour)
                    hour = timecard['in_time'].hour
                    if hour < day_start.hour:
                        hour += 24

                    in_time_str = str(hour) + ':'\
                        + timecard['in_time'].strftime('%M')
                else:
                    in_time_str = ''

                if 'out_time' in timecard:
                    # 23:00 -> 26:00
                    timecard['out_time'] += datetime.timedelta(minutes=day_start.minute)
                    timecard['out_time'] += datetime.timedelta(hours=day_start.hour)
                    hour = timecard['out_time'].hour
                    if hour < day_start.hour:
                        hour += 24
                    out_time_str = str(hour) + ':'\
                        + timecard['out_time'].strftime('%M')
                else:
                    out_time_str = ''

                writer.writerow([
                    date_str, in_time_str, out_time_str, str(timecard['working_hours']), '', ''])

            writer.writerow(['', '', '', '', '', ''])


def main():
    commandline_vars = option_parser()
    data_list, encode = input_data(commandline_vars['input_filename'])
    shaped_data = data_shaping(data_list, commandline_vars['period'])
    if commandline_vars['format_num'] == 0:
        output_data0(commandline_vars['output_filename'], encode, shaped_data)
    elif commandline_vars['format_num'] == 1:
        output_data1(commandline_vars['output_filename'], encode, shaped_data)
  
if __name__ == '__main__':
    main()
